tools/base/fpgad/plugins/fpgad-dc/fpgad-dc.c

```diff
@@ -0,0 +1,937 @@
+			}
+			res = false;
+		} else if (HIGH_WARN(s) || LOW_WARN(s)) {
+			opae_api_send_EVENT_POWER_THERMAL(dc->base_device);
+			BIT_SET_SET(dc->state_tripped, s->id);
+			if (!BIT_IS_SET(dc->state_last, s->id)) {
+				LOG("sensor '%s' warning.\n", s->name);
+			}
+		}
+
+		if (HIGH_NORMAL(s) || LOW_NORMAL(s)) {
+			if (BIT_IS_SET(dc->state_last, s->id)) {
+				negative_trans = true;
+				LOG("sensor '%s' back to normal.\n", s->name);
+			}
+		}
+
+		if (BIT_IS_SET(dc->state_last, s->id) &&
+		    BIT_IS_SET(dc->state_tripped, s->id)) {
+			LOG_MOD(tripped_count,
+				"sensor '%s' still tripped.\n", s->name);
+		}
+	}
+
+	if (negative_trans) {
+		for (i = 0 ; i < dc->num_sensors ; ++i) {
+			if (BIT_IS_SET(dc->state_tripped, dc->sensors[i].id))
+				break;
+		}
+
+		if (i == dc->num_sensors) {
+			// no remaining tripped sensors
+			tripped_count = 0;
+		}
+	}
+
+	for (i = 0 ; i < dc->num_sensors ; ++i) {
+		dc_sensor *s = &dc->sensors[i];
+		if (BIT_IS_SET(dc->state_tripped, s->id))
+			BIT_SET_SET(dc->state_last, s->id);
+		else
+			BIT_SET_CLR(dc->state_last, s->id);
+	}
+
+	memset_s(dc->state_tripped, (dc->max_sensor_id + 7) / 8, 0);
+
+	return res;
+}
+
+STATIC fpga_result dc_unload_driver(dc_device *dc)
+{
+	fpga_token token;
+	fpga_result res;
+	fpga_properties prop = NULL;
+	char path[SYSFS_PATH_MAX];
+	uint16_t seg = 0;
+	uint8_t bus = 0;
+	uint8_t dev = 0;
+	uint8_t fn = 0;
+
+	token = dc->base_device->token;
+
+	res = fpgaGetProperties(token, &prop);
+	if (res != FPGA_OK) {
+		LOG("failed to get fpga properties.\n");
+		return res;
+	}
+
+	if ((fpgaPropertiesGetSegment(prop, &seg) != FPGA_OK) ||
+	    (fpgaPropertiesGetBus(prop, &bus) != FPGA_OK) ||
+	    (fpgaPropertiesGetDevice(prop, &dev) != FPGA_OK) ||
+	    (fpgaPropertiesGetFunction(prop, &fn) != FPGA_OK)) {
+		LOG("failed to get PCI attributes.\n");
+		fpgaDestroyProperties(&prop);
+		return res;
+	}
+
+	fpgaDestroyProperties(&prop);
+
+	//           11111111112222222222333
+	// 012345678901234567890123456789012
+	// /sys/bus/pci/devices/ssss:bb:dd.f
+
+	snprintf_s_i(path, SYSFS_PATH_MAX,
+		     "/sys/bus/pci/devices/%04x:",
+		     (int)seg);
+
+	snprintf_s_i(&path[26], SYSFS_PATH_MAX - 26,
+		     "%02x:", (int)bus);
+
+	snprintf_s_i(&path[29], SYSFS_PATH_MAX - 29,
+		     "%02x.", (int)dev);
+
+	snprintf_s_i(&path[32], SYSFS_PATH_MAX - 32,
+		     "%d/remove", (int)fn);
+
+	LOG("writing 1 to %s to remove driver.\n", path);
+	printf("writing 1 to %s to remove driver.\n", path);
+
+	if (file_write_string(path, "1\n", 2)) {
+		LOG("failed to write \"%s\"\n", path);
+		return FPGA_EXCEPTION;
+	}
+	return FPGA_OK;
+}
+
+STATIC fpga_result dc_force_pci_rescan(void)
+{
+	const char *path = "/sys/bus/pci/rescan";
+	LOG("writing 1 to %s to rescan the device.\n", path);
+	if (file_write_string(path, "1\n", 2)) {
+		LOG("failed to write \"%s\"\n", path);
+		return FPGA_EXCEPTION;
+	}
+	return FPGA_OK;
+}
+
+STATIC pthread_mutex_t cool_down_lock = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;
```

Comment:
I've removed the cool_down_lock from the VC plugin.  Should probably do the same here.